[docs] Update the TypeScript guide, describe the 'compiler options' feature (#5664)
DevExpress/testcafe: docs/articles/documentation/guides/concepts/typescript-and-coffeescript.md

TypeScript comes with rich coding assistance,
painless scalability, check-as-you-type code verification, and much more.

> You do not need to manually compile TypeScript tests. TestCafe does this automatically on launch.

TestCafe bundles the TypeScript declaration file with the npm package, so you do not need to install it separately.

### Write Tests with TypeScript

To start writing tests with TypeScript, install TestCafe into your project directory. For more information, see [Install TestCafe](../basic-guides/install-testcafe.md#local-installation).

Import the`testcafe` module in the beginning of each TypeScript file:

```js
import { Selector } from 'testcafe';
```

Text editors with code completion capabilities (e.g., VS Code, Sublime Text, WebStorm, etc.) load the TestCafe declaration file when the testcafe module is imported.

![Writing Tests with TypeScript](../../../images/typescript-support.png)

> If installed [globally](../basic-guides/install-testcafe.md#global-installation), TestCafe will successfully compile and run your tests written in TypeScript. However, your IDE will not be able to find the TestCafe declaration file and provide code completion.

> The TestCafe syntax is identical for both JavaScript and TypeScript. 

Reviewer: ```js
The test syntax is identical for both JavaScript and TypeScript.
The test execution reports include information about TypeScript compilation errors if there are any.
```

> Using TypeScript brings you all the advantages of strongly typed languages: rich coding assistance,

Reviewer 2: ```suggestion
TypeScript comes with rich coding assistance,
```

> painless scalability, check-as-you-type code verification and much more.

Reviewer 2: ```suggestion
painless scalability, check-as-you-type code verification, and much more.
```

> TestCafe automatically compiles TypeScript tests before launching them, so you do not need to compile the TypeScript code.

TestCafe bundles the TypeScript declaration file with the npm package, so you do not need to install it separately.

### Write Tests with TypeScript

To start writing tests with TypeScript, install TestCafe into your project directory. For more information, see [Install TestCafe](../basic-guides/install-testcafe.md#local-installation).

> Import the testcafe module in the beginning of each TypeScript file:

Reviewer: Add quotes:

```js
Import the `testcafe` module...
```

```js
import { Selector } from 'testcafe';
```

> Text editors with code completion capabilities (e.g. VS Code, Sublime Text, WebStorm, etc.) load the TestCafe declaration file when you import the testcafe module.

Reviewer 2: ```suggestion
Text editors with code completion capabilities (e.g., VS Code, Sublime Text, WebStorm, etc.) load the TestCafe declaration file when the testcafe module is imported.
```

![Writing Tests with TypeScript](../../../images/typescript-support.png)

> If installed [globally](../basic-guides/install-testcafe.md#global-installation), TestCafe will successfully compile and run your tests written in TypeScript. However, your IDE will not be able to find the TestCafe declaration file and provide code completion.

The TestCafe syntax is identical for both JavaScript and TypeScript.

> When you run a TypeScript test, Testcafe outputs the compilation errors that it encounters.

Reviewer 2: outputs?

### Customize Compiler Options

TestCafe users can modify [the settings](https://www.typescriptlang.org/docs/handbook/compiler-options.html) of the TypeScript compiler in one of the following three ways:

* the [--compiler-options](../../reference/command-line-interface.md#--compiler-options) command line parameter,

    ```sh
    testcafe chrome my-tests --compiler-options

>     typescript.options.lib=ES5,WebWorker;typescript.typesRoot='this value contains spaces'

Reviewer 3: ```suggestion
    typescript.options.lib=lib.es5.d.ts,lib.webworker.d.ts;typescript.typesRoot='this value contains spaces'
```

### Customize Compiler Options

> TestCafe users can modify the settings of the [TypeScript compiler](https://www.typescriptlang.org/docs/handbook/compiler-options.html) to enable features such as JSX compilation, and support for React type aliases.

Reviewer: `JSX compilation` is enabled by default. Remove this phrase.

> TestCafe users can modify the settings of the [TypeScript compiler](https://www.typescriptlang.org/docs/handbook/compiler-options.html) to enable features such as JSX compilation, and support for React type aliases.

Reviewer: When you say the `modify`, you need to show here the option that can be modified.
So, add link to the modified settings:

```js
TestCafe users can modify the [settings](link-to-the-default-typescript-compilation-settings)
```

Reviewer: Add a note

```js
At present, you can customize options only for TypeScript compiler.
```

Author: AFAIK we do not 'advertise' future features in documentation, so there's no need to phrase it like that. Right now this compiler setting feature does not come across as a generic because the article only ever mentions TypeScript. I think that users are only going to be more confused if they see a statement such as the one you propose.

Reviewer: The 'compiler options' feature is a generic interface to specify compiler options for all supported compilers: `es-next`, `typescript` and `flow`. At present, the customer can specify options only for TypeScript compiler. In the future, this method will be used to specify options for other compilers.

@VasilyStrelyaev

Reviewer 4: True, but this is the `typescript-and-coffeescript.md` topic, so I don't think that we ought to say about other compilers.
When we support more, we'll have a different guide about compilers in general.

Although, what Mikhail says applies to the reference. It's good to introduce the `compilerOptions` flag/property as the place where we can set the compiler options in general, and then note that it currently supports TypeScript.

Use one of the following three ways to modify the compiler settings:

* the [--compiler-options](../../reference/command-line-interface.md#--compiler-options) command line parameter,

    ```sh
    testcafe chrome my-tests --compiler-options 
    typescript.options.lib=ES5,WebWorker;typescript.typesRoot='this value contains spaces'
    ```

* the [runner.compilerOptions](../../reference/testcafe-api/runner/compileroptions.md) API method,

    ```js

>     runner.compilerOptions([

Reviewer: I wrote the wrong example.
The correct one is

```js
runner.compilerOptions({
    "typescript": {
        customCompilerModulePath: '../node_modules/typescript-v4',
     }});
```
